### PR TITLE
Refactoring/223 improve integration tests

### LINKED ISSUE #223: Improve integration tests
## Situation

The integration tests in its current format have a few problems which we need discuss and resolve:
-tests are too complicated to run even with the documentation;
-tests' data contains python code;
-tests are slow;
-test coverage is not enough: we have 14 dialects, only 4 of them are covered with integration tests and only two of these tests are run in Travis;
-it's not documented how to write your own integration test.

## Additional sub-tasks:

1) Test type mapping for tables with invalid Impala Types to TeradataSqlDialectIT class.
2) Add datatype tests to SybaseSqlDialectIT class.
3) Add tests for data types and test Expanding of SELECT * if elements of select list require casting in OracleSqlDialectIT class
4) Test type mapping for tables with invalid Impala Types in ImpalaSqlDialectIT test class public void testTypeMapping() method.

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**
Closes #223

### CODE REVIEW

File: src/test/java/com/exasol/matcher/ResultSetMatcher.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        return doesObjectMatches("Integer", expected, actual);
+    }
+
+    private boolean doesStringMatch(final ResultSet actualRow, final int column) throws SQLException {
+        final String expected = this.expectedResultSet.getString(column);
+        final String actual = actualRow.getString(column);
+        return doesObjectMatches("String", expected, actual);
+    }
+
+    private <T> boolean doesObjectMatches(final String dataTypeName, final T expectedValue, final T actualValue) {
```
Comment (redcatbear) [inline src/test/java/com/exasol/matcher/ResultSetMatcher.java]: Rename to `doesObjectMatch`.

File: src/test/java/com/exasol/matcher/ResultSetMatcher.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        return doesObjectMatches("BigDecimal", expected, actual);
+    }
+
+    private boolean doesBooleanMatch(final ResultSet actualRow, final int column) throws SQLException {
+        final boolean expected = this.expectedResultSet.getBoolean(column);
+        final boolean actual = actualRow.getBoolean(column);
+        return doesPrimitiveTypeMatches("Boolean", expected, actual);
+    }
+
+    private <T> boolean doesPrimitiveTypeMatches(final String dataTypeName, final T expectedValue,
```
Comment (redcatbear) [inline src/test/java/com/exasol/matcher/ResultSetMatcher.java]: Rename to `doesPrimitiveTypeMatch`.

File: pom.xml [hunk 3]
```diff
@@ -154,6 +154,28 @@
             <artifactId>virtual-schema-common-jdbc</artifactId>
             <version>${vscjdbc.version}</version>
         </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply pom.xml]: I think we found the problem. It should be solved after the release of containers on maven

File: integration-test-data/integration-test-travis.yaml [hunk 4]
```diff
@@ -8,7 +8,7 @@ general:
   additionalJDBCDriverDir: /vagrant/drivers/
 
 exasol:
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply integration-test-data/integration-test-travis.yaml]: done

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 5]
```diff
@@ -1,476 +1,762 @@
...
+
+    private static final String SCHEMA_TEST_MIXED_CASE = "SCHEMA_TEST_Mixed_Case";
+    private static final String TABLE_MIXED_CASE = "Table_Mixed_Case";
+    private static final String VIRTUAL_SCHEMA_EXA_MIXED_CASE = "VIRTUAL_SCHEMA_EXA_Mixed_Case";
+
+    @Container
+    private static final ExasolContainer<? extends ExasolContainer<?>> container = new ExasolContainer<>(
+            ExasolContainerConstants.EXASOL_DOCKER_IMAGE_REFERENCE) //
+                    .withLogConsumer(new Slf4jLogConsumer(LOGGER)) //
+                    .withClusterLogsPath(Path.of("target/mylogs"));
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: done

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 6]
```diff
@@ -1,476 +1,762 @@
...
+    private static final String TABLE_ALL_EXASOL_DATA_TYPES = "TABLE_ALL_EXASOL_TYPES";
+    private static final String TABLE_WITH_NULLS = "TABLE_WITH_NULLS";
+    private static final String TABLE_SIMPLE_VALUES = "TABLE_SIMPLE_VALUES";
+    private static final String TABLE_JOIN_1 = "TABLE_JOIN_1";
+    private static final String TABLE_JOIN_2 = "TABLE_JOIN_2";
+    private static final String VIRTUAL_SCHEMA_JDBC = "VIRTUAL_SCHEMA_JDBC";
+    private static final String VIRTUAL_SCHEMA_JDBC_LOCAL = "VIRTUAL_SCHEMA_JDBC_LOCAL";
+    private static final String VIRTUAL_SCHEMA_EXA = "VIRTUAL_SCHEMA_EXA";
+    private static final String VIRTUAL_SCHEMA_EXA_LOCAL = "VIRTUAL_SCHEMA_EXA_LOCAL";
+
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: done

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 7]
```diff
@@ -1,476 +1,762 @@
...
+    private static final String SCHEMA_TEST_MIXED_CASE = "SCHEMA_TEST_Mixed_Case";
+    private static final String TABLE_MIXED_CASE = "Table_Mixed_Case";
+    private static final String VIRTUAL_SCHEMA_EXA_MIXED_CASE = "VIRTUAL_SCHEMA_EXA_Mixed_Case";
+
+    @Container
+    private static final ExasolContainer<? extends ExasolContainer<?>> container = new ExasolContainer<>(
+            ExasolContainerConstants.EXASOL_DOCKER_IMAGE_REFERENCE) //
+                    .withLogConsumer(new Slf4jLogConsumer(LOGGER)) //
+                    .withClusterLogsPath(Path.of("target/mylogs"));
+    private static Statement statement;
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: We discussed it and decided not to fix

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 8]
```diff
@@ -1,476 +1,762 @@
...
-        stmt.execute("CREATE TABLE \"Table_Mixed_Case\" (\"Column1\" int, \"column2\" int, COLUMN3 int)");
-        stmt.execute("INSERT INTO \"Table_Mixed_Case\" VALUES (1, 2, 3)");
+    static void beforeAll() throws SQLException, BucketAccessException, InterruptedException, TimeoutException {
+        TimeUnit.SECONDS.sleep(20); // TODO need to be fixed in the test-containers
+        final Bucket bucket = container.getDefaultBucket();
+        final Path pathToRls = Path.of("target/" + VIRTUAL_SCHEMAS_JAR_NAME_AND_VERSION);
+        bucket.uploadFile(pathToRls, VIRTUAL_SCHEMAS_JAR_NAME_AND_VERSION);
+        final Connection connection = container.createConnectionForUser(container.getUsername(),
+                container.getPassword());
+        statement = connection.createStatement();
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: We discussed it and decided not to fix

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 9]
```diff
@@ -1,476 +1,762 @@
...
+    private static void createTestTablesForJoinTests() throws SQLException {
+        statement.execute("CREATE TABLE " + SCHEMA_TEST + "." + TABLE_JOIN_1 + "(x INT, y VARCHAR(100))");
+        statement.execute("INSERT INTO " + SCHEMA_TEST + "." + TABLE_JOIN_1 + " VALUES (1,'aaa'), (2,'bbb')");
+        statement.execute("CREATE TABLE " + SCHEMA_TEST + "." + TABLE_JOIN_2 + "(x INT, y VARCHAR(100))");
+        statement.execute("INSERT INTO " + SCHEMA_TEST + "." + TABLE_JOIN_2 + " VALUES (2,'bbb'), (3,'ccc')");
+    }
+
+    private static void createConnection() throws SQLException {
+        statement.execute("CREATE CONNECTION " + JDBC_EXASOL_CONNECTION + " " //
+                + "TO 'jdbc:exa:localhost:8888' " //
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: We discussed it and decided not to fix it right now

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 10]
```diff
@@ -1,476 +1,761 @@
...
+
+    private void assertSelectColumnsResult(final String virtualSchemaName, final String columns) throws SQLException {
+        final ResultSet expected = statement
+                .executeQuery("SELECT " + columns + " FROM " + SCHEMA_TEST + "." + TABLE_ALL_EXASOL_DATA_TYPES);
+        final ResultSet actual = statement
+                .executeQuery("SELECT " + columns + " FROM " + virtualSchemaName + "." + TABLE_ALL_EXASOL_DATA_TYPES);
+        assertThat(actual, matchesResultSet(expected));
+    }
+
+    // TODO: This test excludes VIRTUAL_SCHEMA_EXA virtual schema because of the wrong data mapping caused by driver.
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: done

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 11]
```diff
@@ -1,476 +1,761 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("getVirtualSchemaVariantsAll")
+    void testTimestamp(final String virtualSchemaName) throws SQLException {
+        assertSelectColumnsResult(virtualSchemaName, "C10, C11");
+    }
+
+    // TODO: This test excludes VIRTUAL_SCHEMA_JDBC and VIRTUAL_SCHEMA_EXA virtual schema because of the bug on the
+    // loader side.
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: done

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java [hunk 12]
```diff
@@ -1,476 +1,761 @@
...
-        assertNextRow(result, Short.valueOf("1"));
-        matchSingleRowExplain(query,
-                "IMPORT FROM EXA AT 'localhost:8888' USER 'sys' IDENTIFIED BY 'exasol' STATEMENT 'SELECT 1 FROM \"NATIVE_EXA_IT\".\"SIMPLE_VALUES\"'",
-                IS_LOCAL);
+    @ParameterizedTest
+    @MethodSource("getVirtualSchemaVariantsExa")
+    void testGroupConcatOrderByDescNullsLast(final String virtualSchemaName) {
+        final String query = "SELECT GROUP_CONCAT(A ORDER BY C DESC NULLS LAST) FROM " + virtualSchemaName + "."
+                + TABLE_SIMPLE_VALUES;
+        assertAll(() -> assertExpressionExecutionStringResult(query, "3,2,1,3,2,1"), //
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: Only explain virtual is different for this test... Should I remove it?

File: src/test/java/com/exasol/matcher/ResultSetMatcher.java (hunk 1 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/matcher/ResultSetMatcher.java]: fixed

File: src/test/java/com/exasol/matcher/ResultSetMatcher.java (hunk 2 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/matcher/ResultSetMatcher.java]: done

File: src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/test/java/com/exasol/adapter/dialects/exasol/ExasolSqlDialectIT.java]: I didn't see that one, you are right.

**Pull request merged**